🤝  [19834] Add dapp connection flow with a basic design

shivekkhurana opened the pull request:
### Summary
This is an ongoing effort to implement wallet connect. This PR connects the existing methods to the universal qr code scanner. In case a valid Wallet Connect QR is scanned, the user gets an option to accept the connection.

In case the QR code is expired or the wallet connect version is not 2, the user gets a relevant error toast.

### Review notes

- Hard coded the first account
- The data item component doesn't support networks or accounts yet
- The quo category component cannot show a data-item yet
- The above two points mean that networks or accounts cannot be changed
- I didn't add the validation to ensure we support required namespaces
- Connected accept and decline button

#### Platforms
<!-- (Optional. Specify which platforms should be tested) -->

- Android
- iOS

#### Areas that maybe impacted
- Global QR scanner

### Steps to test
- Goto https://lab.web3modal.com/library/ethers-siwe/
- Select the mainnet in networks (this is important)
- Click Connect Wallet, a QR code will pop up
- Use the universal scanner to scan that QR code, you'll see a screen to accept or reject the session

status: ready

shivekkhurana committed:
🔳 QR on success not being called

- IDK why, trying things out

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | ec6639153d90b9306fbc753aadeeb107b4b642cb | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-20204/1/) | 2024-05-28 09:58:05 | ~6 min | `tests` | [:page_facing_up:`log`](https://status-im-mobile-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-240528-095151-ec6639153d90b9306fbc753aadeeb107b4b642cb-pr20204-tests.log) |

clauxx commented on src/status_im/contexts/shell/qr_reader/view.cljs:
```diff
@@ -68,6 +69,34 @@
    [:toasts/upsert invalid-qr-toast]
    300))
 
+(defn- handle-wallet-connect
+  [scanned-text]
+  (let [parsed-uri         (wallet-connect/parse-uri scanned-text)
+        version            (-> parsed-uri :version)
+        expired?           (wc-utils/timestamp-expired? (-> parsed-uri :expiryTimestamp))
+        version-supported? (wc-utils/version-supported? version)]
+
+    (cond
+      expired?
+      (debounce/debounce-and-dispatch
+       [:toasts/upsert
+        {:type  :negative
+         :theme :dark
+         :text  (i18n/label :t/wallet-connect-qr-expired)}]
+       300)
+
+      (not version-supported?)
+      (debounce/debounce-and-dispatch
+       [:toasts/upsert
+        {:type  :negative
+         :theme :dark
+         :text  (i18n/label :t/wallet-connect-version-not-supported
+                            {:version version})}]
+       300)
+
+      :else
+      (debounce/debounce-and-dispatch [:wallet-connect/pair scanned-text] 300))))
+
```
this should be an event

clauxx commented on src/status_im/contexts/wallet/common/scan_account/view.cljs:
```diff
@@ -29,10 +30,12 @@
-      :validate-fn     #(contains-supported-address? %)
+      :validate-fn     (fn [scanned-text]
+                         (contains-supported-address? scanned-text))
       :on-success-scan (fn [result]
                          (let [address (extract-address result)]
                            (when on-result (on-result address))
-                           (debounce/debounce-and-dispatch
-                            [:wallet/scan-address-success address]
-                            300)))}]))
+                           (when address
+                             (debounce/debounce-and-dispatch
+                              [:wallet/scan-address-success address]
+                              300))))}]))
```
is the WC part handled here?

shivekkhurana replied:
No, this is some other QR code component. I mistook it for the global reader earlier.

clauxx replied:
I guess these changes should be removed then

clauxx commented on src/status_im/contexts/wallet/wallet_connect/events.cljs:
```diff
@@ -36,7 +36,9 @@
  (fn [{:keys [db]} [proposal]]
-   {:db (assoc db :wallet-connect/current-proposal proposal)}))
+   {:db            (assoc db :wallet-connect/current-proposal proposal)
+    :open-modal-fx [:screen/wallet.wallet-connect-session-proposal
+                    (:theme db)]}))
```
effects should go into the `:fx` key e.g. `{:db ... :fx [[:open-modal-fx [...]]]}`

clauxx replied:
Also, I don't really see the effect being used directly in the codebase. Usually we're dispatching the `:open-modal` event, which does the same and also takes care of injecting the theme.

shivekkhurana replied:
Yeah, but in this case, we cannot dispatch the event (to open modal), because this event is not user generated. 

So I did what the `open-modal` event handler does internally.

clauxx replied:
why does it matter that it's not user generated?

ulisesmac replied:
@shivekkhurana 
What @clauxx is referring to is to use the `:fx` effect to dispatch re-frame events

shivekkhurana replied:
I understand the `:fx` thing. My answer was for "Usually we're dispatching the :open-modal". We cannot `rf/dispatch` here, but I'll move `:open-modal-fx` to `:fx` key.

clauxx replied:
> We cannot rf/dispatch here

I don't understand why we can't. Couldn't we just do 

```clj
{:db (...)
 :fx [[:dispatch [:open-modal :screen/wallet.wallet-connect-session-proposal]]]}
```

shivekkhurana replied:
> > We cannot rf/dispatch here
> 
> I don't understand why we can't. Couldn't we just do
> 
> ```clojure
> {:db (...)
>  :fx [[:dispatch [:open-modal :screen/wallet.wallet-connect-session-proposal]]]}
> ```

I understand now. I didn't know this is possible. I'll make the required change.

clauxx commented on src/status_im/contexts/wallet/wallet_connect/session_proposal/style.cljs:
```diff
@@ -0,0 +1,34 @@
+(ns status-im.contexts.wallet.wallet-connect.session-proposal.style
+  (:require [quo.foundations.colors :as colors]))
+
+(def container
+  {})
+
```
if not used, should be removed

clauxx commented on src/status_im/contexts/wallet/wallet_connect/session_proposal/view.cljs:
```diff
@@ -0,0 +1,123 @@
+(defn- dapp-metadata
+  []
+  (let [proposer  (rf/sub [:wallet-connect/session-proposer])
+        metadata  (-> proposer :metadata)
+        icon      (-> metadata :icons first)
+        dapp-name (-> metadata :name)
+        url       (-> metadata :url)]
```
 in most of those threading is not needed

alwx replied:
you could simply destructure here, right at line 14

clauxx commented on src/status_im/contexts/wallet/wallet_connect/session_proposal/view.cljs:
```diff
@@ -0,0 +1,123 @@
+        labels    [(i18n/label :t/check-your-account-balance-and-activity)
+                   (i18n/label :t/request-txns-and-message-signing)]]
+    [rn/view {:style style/approval-note-container}
+     [rn/text {:style style/approval-note-title}
+      (i18n/label :t/dapp-will-be-able-to {:dapp-name dapp-name})]
+     (map-indexed
+      (fn [idx label]
+        ^{:key (str idx label)}
+        [rn/view {:style style/approval-note-li}
+         [quo/icon :i/bullet
+          {:color colors/neutral-50}]
+         [rn/view {:style style/approval-li-spacer}]
+         [rn/text label]])
```
should we use quo/text instead of rn/text?

clauxx reviewed (commented):
After testing I noticed that once you press connect and it's successful, the modal doesn't disappear.

ulisesmac commented on src/status_im/contexts/shell/qr_reader/view.cljs:
```diff
@@ -68,6 +69,34 @@
+(defn- handle-wallet-connect
+  [scanned-text]
+  (let [parsed-uri         (wallet-connect/parse-uri scanned-text)
+        version            (-> parsed-uri :version)
+        expired?           (wc-utils/timestamp-expired? (-> parsed-uri :expiryTimestamp))
```
Related to code style, IMO, having a `->` of only one step is not a good idea, since `(f value)` is clear enough and more simple than  `(-> value f)` .

In `expired?` the thread first could also contain `timestamp-expired?`

shivekkhurana replied:
I missed the thread first for expired, fixed it now. thanks <3

clauxx commented on src/status_im/contexts/wallet/wallet_connect/utils.cljs:
```diff
@@ -0,0 +1,29 @@
+(ns status-im.contexts.wallet.wallet-connect.utils
+  (:require [react-native.wallet-connect :refer [parse-uri]]))
```
should avoid using refer according to the guidelines

clauxx commented on src/status_im/config.cljs:
```diff
@@ -24,7 +24,7 @@
 (goog-define ALCHEMY_OPTIMISM_MAINNET_TOKEN "")
 (goog-define ALCHEMY_OPTIMISM_GOERLI_TOKEN "")
 (goog-define ALCHEMY_OPTIMISM_SEPOLIA_TOKEN "")
-(goog-define WALLET_CONNECT_PROJECT_ID "")
```
could you rebase and push without the project id? I'll add a task now to add it to the secrets so we don't expose it. It's kinda unclear what to do with it, but for the extra safety we should not push it for now

clauxx commented on src/status_im/contexts/wallet/wallet_connect/events.cljs:
```diff
@@ -124,3 +129,32 @@
                                      (rf/dispatch [:wallet-connect/reset-current-session]))}]]})))
+
+(rf/reg-event-fx
+ :wallet-connect/potential-connection-uri-scanned
```
I think we can skip `potential` here for a less verbose name e.g. `:wallet-connect/on-scan-connection`

clauxx commented on src/status_im/contexts/wallet/wallet_connect/events.cljs:
```diff
@@ -124,3 +129,32 @@
+
+       :else
+       {:fx [[:dispatch [:wallet-connect/pair scanned-text]]
+             [:dispatch [:dismiss-modal :screen/wallet.wallet-connect-session-proposal]]]}))))
```
Why do we dismiss the session proposal modal when we scan the qr?

shivekkhurana replied:
Yeah, this should be in `approve-session` event. I'll move it there.

shivekkhurana committed:
🔨 Fix lint

shivekkhurana committed:
🙃 Stupid of me

- My handler was not being called because I wrote the code in a
different QR scanner

shivekkhurana committed:
✅ Approval screen taking shape

shivekkhurana committed:
🧹 Lint fix

shivekkhurana committed:
🎛️ Wallet connect session screen shows up

- Hard coded the first account
- The data item component doesn't support networks or accounts yet
- The quo category component cannot show a data-item yet
- Connected accept and decline button

shivekkhurana committed:
🧰 Fix review issues

shivekkhurana committed:
🔧 Rename event and move dispatch

shivekkhurana committed:
🔧 Fix lint
